### PR TITLE
Offset overlapping vehicles horizontally

### PR DESCRIPTION
**Pull request by arkadyan**
Asana ticket: [Offset overlapping vehicles horizontally](https://app.asana.com/0/1112935048846093/1123838037403440)

### CODE REVIEW

**Comment by arkadyan** [comment]
Screen Shot 2019-06-10 at 11 41 49

File: assets/src/models/ladderVehiclePosition.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  const gappedLanes: number[] = range(occupiedLanes.length).filter(
+    lane => !occupiedLanes.includes(lane)
+  )
+  if (gappedLanes && gappedLanes.length) {
+    return gappedLanes[0]
+  }
+
+  return Math.max(...occupiedLanes) + 1
```
Comment (skyqrose) [inline assets/src/models/ladderVehiclePosition.ts]: Could there be a simpler way to implement this by sorting the input and walking through it til you find the first gap?

File: assets/src/models/ladderVehiclePosition.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+export const byDirectionAndY = (
+  a: LadderVehiclePosition,
+  b: LadderVehiclePosition
+): number => {
+  if (a.vehicleDirection === VehicleDirection.Down) {
+    if (b.vehicleDirection === VehicleDirection.Up) {
+      return -1
+    } else {
+      if (a.y > b.y) {
+        return -1
```
Comment (skyqrose) [inline assets/src/models/ladderVehiclePosition.ts]: This comparison is a little hard to follow. It could use a quick human description "Down is less than Up. If they're the same direction the vehicle behind is less than the vehicle in front" or something similar (idk if I got the directions right)

What impact does the order of the sorting have? If it's important, that should probably get noted too.

File: assets/src/models/ladderVehiclePosition.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        )
+        const lane = firstOpenLane(occupiedLanes)
+
+        return [
+          {
+            ...ladderVehiclePosition,
+            lane,
+          },
+          ...previousLadderVehiclePositions,
+        ]
```
Comment (skyqrose) [inline assets/src/models/ladderVehiclePosition.ts]: Does this reduce flip the order of the input? Is it important if it does or not?

File: assets/src/models/ladderVehiclePosition.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    areOverlapping(previous, ladderVehiclePosition)
+  )
+
+export const areOverlapping = (
+  a: LadderVehiclePosition,
+  b: LadderVehiclePosition
+): boolean => {
+  return (
+    a.vehicleDirection === b.vehicleDirection &&
+    areOverlappingRanges(yRange(a), yRange(b))
```
Comment (skyqrose) [inline assets/src/models/ladderVehiclePosition.ts]: This could just be `abs(a.y - b.y) < minDistanceBetween`, and you wouldn't need the range stuff.

File: assets/src/models/ladderVehiclePosition.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+export const xInLaneWithOffset = (ladderVehicleHorizontalOffset: number) => ({
+  x,
+  vehicleDirection,
+  lane,
+}: LadderVehiclePosition): number => {
+  const laneOffset = (lane || 0) * ladderVehicleHorizontalOffset
+  return vehicleDirection === VehicleDirection.Up
+    ? x + laneOffset
+    : x - laneOffset
```
Comment (skyqrose) [inline assets/src/models/ladderVehiclePosition.ts]: I feel like this could be simpler. Like, maybe `x` doesn't need to be part of `LadderVehiclePosition`, and you could do all the x calculations at once based on the `lane` and `vehicleDirection`.

File: assets/src/components/ladder.tsx [hunk 6]
```diff
@@ -60,18 +64,77 @@ const Ladder = ({
...
+        timepointStatusY(
+          vehicle.timepoint_status,
+          timepoints,
+          timepointSpacingY,
+          vehicleDirection
+        ) || -10
+
+      return {
+        vehicle,
+        vehicleDirection,
```
Comment (skyqrose) [inline assets/src/components/ladder.tsx]: There's some interesting algorithm stuff, but it's pretty tied up in the vehicle stuff. Do you think it'd be possible to split apart? Maybe something like
`putIntoLanes = (input: [{y: number, payload: whatever}], minSpaceBetween: number): [{y: number, payload: whatever, lane: number}]`
Maybe you could partition the up and down vehicle and call it separately on the two sides so that the algorithmic stuff doesn't need to think about `VehicleDirection` at all? I think it'd end up being easier to reason about and test if the lane algorithm didn't need to think about vehicle stuff, and if all the vehicle logic didn't have any lane algorithm mixed in.

File: assets/src/components/ladderVehicle.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+  return (
+    <g
+      className={`m-ladder__vehicle ${selectedClass}`}
+      transform={`translate(${xInLane(ladderVehiclePosition)},${y -
+        centerOfVehicleGroupY}) rotate(${rotation},${centerOfVehicleGroupX},${centerOfVehicleGroupY})`}
+      onClick={() => dispatch(selectVehicle(vehicle.id))}
+    >
+      <VehicleIcon scale={0.625} />
+      <VehicleLabel vehicle={vehicle} rotation={rotation} x={0} y={26} />
```
Comment (skyqrose) [inline assets/src/components/ladderVehicle.tsx]: #72 pulls the label inside the `VehicleIcon`, so this will change quite a bit.

File: assets/src/components/ladder.tsx [hunk 8]
```diff
@@ -1,56 +1,60 @@
...
+  const allLanes: number[] = ladderVehiclePositions.map(
+    ({ lane }: LadderVehiclePosition) => lane || 0
+  )
+  const furthestLane = allLanes.length ? Math.max(...allLanes) : 0
+  return furthestLane + 1
+}
+
+const ladderVehicleHorizontalOffsetForLanes = (maxOccupiedLane: number) => {
+  const cappedLanes = Math.min(maxOccupiedLane, 5)
+  return ((8 - cappedLanes) / 8) * widthOfVehicleGroup
```
[the remark this answers is not in the document]
Comment (arkadyan) [reply assets/src/components/ladder.tsx]: This is defining the offset for each lane based on the max number of vehicles wide this ladder gets in order to condense the vehicles more as there are more of them across. So, if there are 2 vehicles across it will space them at a wider 75% of the icon width, while at 5 vehicles across it'll be tighter at 37.5%. It is capped at that condensed to ensure you can still click on the edge of any icon.

File: assets/src/models/ladderVehiclePosition.ts (hunk 1 above)
Comment (arkadyan) [reply assets/src/models/ladderVehiclePosition.ts]: That was actually my first implementation. It ended up being much less simple because their might not be a gap, or might not be any occuied lanes. This is much more straightforward and these list sizes are never going to be that large.

File: assets/src/models/ladderVehiclePosition.ts (hunk 3 above)
Comment (arkadyan) [reply assets/src/models/ladderVehiclePosition.ts]: It does, but it is not important - the sorting only mattered for the logic in that reduce function, nowhere externally.

File: assets/src/models/ladderVehiclePosition.ts (hunk 4 above)
Comment (arkadyan) [reply assets/src/models/ladderVehiclePosition.ts]: Nice!

File: assets/src/components/ladder.tsx (hunk 6 above)
Comment (arkadyan) [reply assets/src/components/ladder.tsx]: I think if I understand what you are describing it gets to why I wanted to pull out a view model in the first place, to get all the logic out of the components. I think the problem is that this function here doesn't belong here, it belongs in `ladderVehiclePosition`. What if we passed `vehicles` and `timepointStatusY` into a function in `ladderVehiclePosition` that is basically this function, but moving the `x` calculation to aftert we figure out the lanes? I can do that work to show it more concretely, but does that sound like it would be moving in a better direction?

File: assets/src/models/ladderVehiclePosition.ts (hunk 1 above)
Comment (skyqrose) [reply assets/src/models/ladderVehiclePosition.ts]: What about:
```
const sortedOccupiedLanes = occupiedLanes.slice().sort()
for (var i = 0; i < sortedOccupiedLanes.length; i++) { 
  if (sortedOccupiedLanes[i] !== i) {
    return i
  }
}
return sortedOccupiedLanes.length
```
It's not very functional, but I think it's still simpler, and should cover if the list is empty (do the loop 0 times, then return 0), and full (do the loop n times, then return n).

File: assets/src/components/ladder.tsx (hunk 6 above)
Comment (skyqrose) [reply assets/src/components/ladder.tsx]: I think the view model helps a lot, and moving the `x` calculation to after we figure out the lanes would make that even better.

I think we could also separate the algorithmic part of `putIntoLanes` from the view model. Like, `putIntoLanes`, `overlappingPreviousVehicles`, `firstOpenLane`, and friends don't actually need any information about the vehicle, just its y, so could be pulled into a separate file.

There's a good abstraction boundary there that would make it easier to see the view model and how it's made and used without having to wade through any of the algorithmic stuff.

File: assets/src/models/ladderVehiclePosition.ts (hunk 1 above)
Comment (arkadyan) [reply assets/src/models/ladderVehiclePosition.ts]: That works.

File: assets/src/components/ladder.tsx (hunk 6 above)
Comment (arkadyan) [reply assets/src/components/ladder.tsx]: Take a fresh look: I extracted all that logic out of the Ladder component and made the internal logic much more pure about what data it requires (with internal interfaces to demonstrate that).

Comment (arkadyan) [reply assets/src/components/ladder.tsx]: One small additional update: I curried some timepointStatusY variables to avoid passing them around unnecessarily.

**Comment by arkadyan** [comment]
@skyqrose I don't see why this isn't worth a final review at this point. I haven't seen the new ladder width calculation given 4 or 5 vehicles yet, but that's also trivial to tweak later if we discover any issues.

File: assets/src/components/ladder.tsx [hunk 9]
```diff
@@ -1,28 +1,28 @@
...
-export const vehicleDirectionOnLadder = (
-  vehicle: Vehicle,
-  ladderDirection: LadderDirection
-): VehicleDirection =>
-  (vehicle.directionId === 1) ===
-  (ladderDirection === LadderDirection.ZeroToOne)
-    ? VehicleDirection.Down
-    : VehicleDirection.Up
+export const widthOfVehicleGroup = 32
+export const heightOfVehicleGroup = 34
```
Comment (skyqrose) [inline assets/src/components/ladder.tsx]: These are only used in `ladderVehicle.ts`, and could be moved there so they can be private.

File: assets/src/models/ladderVehicle.ts [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+  )
+
+  const vehiclesInLane: VehicleInLane[] = putIntoLanes(vehiclesOnLadder)
+
+  const maxOccupiedLane: number = numOccupiedLanes(vehiclesInLane)
+  const ladderVehicleHorizontalOffset: number = horizontalOffsetForLanes(
+    maxOccupiedLane
+  )
+  const xInLane = xInLaneWithOffset(ladderVehicleHorizontalOffset)
+  // const widthOfLanes = maxOccupiedLane * ladderVehicleHorizontalOffset
```
Comment (skyqrose) [inline assets/src/models/ladderVehicle.ts]: Dead code

File: assets/src/components/ladder.tsx [hunk 11]
```diff
@@ -60,18 +51,69 @@ const Ladder = ({
...
+    timepoints,
+    timepointSpacingY
+  )
+  const { ladderVehicles, widthOfLanes } = ladderVehiclesFromVehicles(
+    vehicles,
+    ladderDirection,
+    timepointStatusY
+  )
+
+  const width = 120 + 2 * widthOfLanes * 1.2
```
Comment (skyqrose) [inline assets/src/components/ladder.tsx]: With the new more descriptive width calculations, this `* 1.2` scaling shouldn't be necessary, assuming we want constant sized margins.

File: assets/src/components/ladder.tsx [hunk 12]
```diff
@@ -60,18 +51,69 @@ const Ladder = ({
...
-          key={vehicle.id}
-          vehicle={vehicle}
-          timepoints={orderedTimepoints}
-          timepointSpacingY={timepointSpacingY}
-          ladderDirection={ladderDirection}
-          selectedVehicleId={selectedVehicleId}
-        />
-      ))}
+    <svg className="m-ladder" height={HEIGHT} width={width} viewBox={viewBox}>
+      {ladderVehicles.map(ladderVehicle => {
```
Comment (skyqrose) [inline assets/src/components/ladder.tsx]: This map function is pretty long, and makes it harder to read the list of everything drawn in the SVG. It could be broken out.

File: assets/src/models/ladderVehicle.ts [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+          ...previousLadderVehiclePositions,
+        ]
+      },
+      []
+    )
+
+const numOccupiedLanes = (vehicles: InLane[]): number => {
+  const allLanes: number[] = vehicles.map(({ lane }) => lane || 0)
+  const furthestLane = allLanes.length ? Math.max(...allLanes) : 0
+  return furthestLane + 1
```
Comment (skyqrose) [inline assets/src/models/ladderVehicle.ts]: `numOccupiedLanes` and `maxOccupiedLanes` get conflated. Since one of the uses subtracts 1 later, I think this function should get renamed `maxOccupiedLane` and the `+ 1` here removed.

The `- 1` when calculated `widthOfLanes` can be removed then. (The `horizontalOffsetForLanes` also uses `maxOccupiedLane`, and might need to tweak its calculations.)

File: assets/tests/models/ladderVehicle.test.ts [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+    ]
+
+    const result = putIntoLanes(original)
+
+    expect(result).toEqual(expected)
+  })
+})
+
+describe("byDirectionAndY", () => {
+  test("sorts the arry by direction, and then by y value", () => {
```
Comment (skyqrose) [inline assets/tests/models/ladderVehicle.test.ts]: typo `arry` -> `array`

File: assets/src/components/ladder.tsx (hunk 12 above)
Comment (arkadyan) [reply assets/src/components/ladder.tsx]: Will be fixed in a future PR
